### PR TITLE
C++: HashCons library

### PR DESCRIPTION
**Pull request by rdmarsh2**
A structural comparison library based on the hashconsing strategy for comparison. The implementation is similar to the existing global value numbering library, plus some tricky handling for function arguments.

### CODE REVIEW

**Comment by rdmarsh2** [comment]
There's a few categories of expressions that aren't covered yet: array literals will need a similar technique to function calls, struct literals may need something more complicated. New and delete can likely be done as an extension of the function call handling. Throw expressions should be simple to implement, as should sizeof, alignof, and company. I haven't decided whether assignments should be hash-consed or not; input would be appreciated.

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+  FunctionCall fc
+) {
+  analyzableNonmemberFunctionCall(fc)
+  or
+  analyzableMemberFunctionCall(fc)
+}
+
+/**
+ * Holds if `fc` is a call to `fcn`, `fc`'s first `i-1` arguments have hash-cons
+ * `list`, and `fc`'s `i`th argument has hash-cons `hc`
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: I think this comment is off by one. Where it says `i-1` it should just say `i`. Take, for example, the case of `i=0`. Then it should say that "`fc`'s first **0** arguments have hash-cons `list`, and `fc`'s **0th** argument has hash-cons `hc`.

Maybe also change "`fc`'s `i`th argument" to "`fc`'s argument at index `i`" or clarify in some other way that the first argument has `i=0`. Also add a full stop at the end of the sentence.

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+  HC_MemberFunctionCall(Function trg, HC qual, HC_Args args) {
+    mk_MemberFunctionCall(trg, qual, args, _)
+  }
+  or
+  // Any expression that is not handled by the cases above is
+  // given a unique number based on the expression itself.
+  HC_Unanalyzable(Expr e) { not analyzableExpr(e,_) }
+
+/** Used to implement hash-consing of argument lists  */
+private cached newtype HC_Args =
```
Comment (xiemaisi) [inline cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: Drive-by comment: When I looked into implementing a similar library for JavaScript, I got a considerable speed-up from adding special-case constructors for argument lists of length one and two. Have you investigated whether this makes a difference for C++?

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+   * Gets an example of an expression with this HC.
+   * This is useful for things like implementing toString().
+   */
+  private Expr exampleExpr() {
+    // Pick the expression with the minimum source location string. This is
+    // just an arbitrary way to pick an expression with this `HC`.
+    result =
+       min(Expr e
+       | this = hashCons(e)
+       | e order by e.getLocation().toString())
```
[the remark this answers is not in the document]
Comment (ian-semmle) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: There should only be 1 Location for any physical location of expressions. There may be another one for the location of statements, and another for the location of other things.

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll [hunk 4]
```diff
@@ -0,0 +1,519 @@
...
+      i > 0
+    )
+    or
+    i = 0 and
+    list = HC_EmptyArgs(fcn)
+  )
+}
+
+/** Gets the hash-cons of expression `e`. */
+cached HC hashCons(Expr e) {
```
[the remark this answers is not in the document]
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: They're in the same stage.

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll (hunk 3 above)
[the remark this answers is not in the document]
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: The results there are very strange; it gives one location for the literal and two for the variable access and the addition. I think that's extractor weirdness, but it might be something in the library

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll [hunk 5]
```diff
@@ -0,0 +1,519 @@
...
+  or
+  HC_Deref(HC p) {
+    mk_Deref(p,_) or
+    mk_PointerFieldAccess(p,_,_) or
+    mk_ImplicitThisFieldAccess_with_qualifier(p,_,_)
+  }
+  or
+  HC_ThisExpr(Function fcn) {
+    mk_ThisExpr(fcn,_) or
+    mk_ImplicitThisFieldAccess(fcn,_,_)
```
[the remark this answers is not in the document]
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: I think that's a holdover from the GVN library. @kevinbackhouse can you confirm?

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll (hunk 3 above)
[the remark this answers is not in the document]
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: to be clear, that seems to be one location for each variable access in the macro expansion, with `loc1 != loc2`, but `loc1.subsumes(loc2)` and `loc2.subsumes(loc1)`

[the remark this answers is not in the document]
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: added a test; I'm still not sure what's going on here. @nickrolfe @ian-semmle can you comment on what's happening in this case?

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll [hunk 6]
```diff
@@ -0,0 +1,519 @@
...
+}
+
+private predicate mk_Deref(
+  HC p, PointerDereferenceExpr deref) {
+  analyzablePointerDereferenceExpr(deref) and
+  p = hashCons(deref.getOperand().getFullyConverted())
+}
+
+private predicate analyzableNonmemberFunctionCall(
+  FunctionCall fc) {
```
[the remark this answers is not in the document]
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: fixed

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll [hunk 7]
```diff
@@ -0,0 +1,519 @@
...
+ * value. The only use for the `HC` of an expression is to find other
+ * expressions that are structurally equal to it. Use the predicate
+ * `hashCons` to get the `HC` for an `Expr`.
+ *
+ * Note: `HC` has `toString` and `getLocation` methods, so that it can be
+ * displayed in a results list. These work by picking an arbitrary
+ * expression with this `HC` and using its `toString` and `getLocation`
+ * methods.
+ */
+class HC extends HCBase {
```
[the remark this answers is not in the document]
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: I think that was for symmetry with GVN from the global value numbering library. I've switched to HashCons for the public class name

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll [hunk 8]
```diff
@@ -0,0 +1,519 @@
...
+  HC() { this instanceof HCBase }
```
[the remark this answers is not in the document]
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: Done

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll [hunk 9]
```diff
@@ -0,0 +1,519 @@
...
+
+private predicate analyzablePointerFieldAccess(PointerFieldAccess access) {
+  strictcount (access.getTarget()) = 1 and
+  strictcount (access.getQualifier().getFullyConverted()) = 1
+}
+
+private predicate mk_PointerFieldAccess(
+  HC qualifier, Field target,
+  PointerFieldAccess access) {
+   analyzablePointerFieldAccess(access) and
```
[the remark this answers is not in the document]
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: fixed

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll [hunk 10]
```diff
@@ -0,0 +1,519 @@
...
+  or
+  exists (float val, Type t
+  | mk_FloatLiteral(val, t, e) and
+    result = HC_FloatLiteral(val, t))
+  or
+  exists (string val, Type t
+  | mk_StringLiteral(val, t, e) and
+    result = HC_StringLiteral(val, t))
+  or
+  // Variable with no SSA information.
```
[the remark this answers is not in the document]
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: done

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll (hunk 1 above)
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: done

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll (hunk 2 above)
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: I haven't. I'll take a look at that, though.

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll (hunk 5 above)
[the remark this answers is not in the document]
Comment (kevinbackhouse) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: Yes, that's correct.

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll (hunk 3 above)
[the remark this answers is not in the document]
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: It does look like the accesses to `y` are duplicated, but somehow there's only one row in the `.expected` file for the `HashCons` object even though I'd expect it to have two locations. So it might be strange, but I suppose it's not a problem.

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll (hunk 5 above)
[the remark this answers is not in the document]
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: It looks like implicit `this` is usually expanded in the extractor, so it may be better to leave that in place and add some notes about it, rather than making the treatment of an implicit `this` be inconsistent.

[the remark this answers is not in the document]
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: It's rare to see an implicit `this` in the AST. I think it happens only for calls to the destructors of fields in compiler-generated destructors. I don't think that's something we need to worry about for HashCons.

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll [hunk 11]
```diff
@@ -0,0 +1,1148 @@
...
+  strictcount(e.getTypeOperand()) = 1
+}
+
+private predicate mk_UuidofOperator(Type t, UuidofOperator e) {
+  analyzableUuidofOperator(e) and
+  t = e.getTypeOperand()
+}
+
+private predicate analyzableTypeidType(TypeidOperator e) {
+  strictcount(e.getAChild()) = 0
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: This comparison is always false as `strictcount` never returns 0.

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll (hunk 5 above)
[the remark this answers is not in the document]
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: I still think it's worth simplifying the hash-consing of field accesses. I don't think it's important that `myField` gets equated with `this.myField` as those two would never occur in the same function because the extractor inserts explicit `this.` in user-written code.

[the remark this answers is not in the document]
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: Yeah, if it's that rare I'll take the logic out;.

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll [hunk 12]
```diff
@@ -0,0 +1,1148 @@
...
+  or
+  HC_MemberFunctionCall(Function trg, HashCons qual, HC_Args args) {
+    mk_MemberFunctionCall(trg, qual, args, _)
+  }
+  or
+  HC_NewExpr(Type t, HC_Alloc alloc, HC_Init init, HC_Align align) {
+    mk_NewExpr(t, alloc, init, align, _, _)
+  }  or
+  HC_NewArrayExpr(Type t, HC_Alloc alloc, HC_Init init, HC_Align align) {
+    mk_NewArrayExpr(t, alloc, init, align, _, _)
```
[the remark this answers is not in the document]
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: no; the type of the newArrayExpr is an ArrayType, so the size is included there.

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll [hunk 13]
```diff
@@ -0,0 +1,1148 @@
...
+  }
+  or HC_ArgCons(HashCons hc, int i, HC_Args list) {
+    mk_ArgCons(hc, i, list, _)
+  }
+
+/**
+ * Used to implement hash-consing of struct initizializers.
+ */
+private newtype HC_Fields =
+  HC_EmptyFields(Class c) {
```
[the remark this answers is not in the document]
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: HC_Fields and HC_Array may need to be separate to handle designated initializers, depending on what decision we make about them

File: cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll (hunk 12 above)
[the remark this answers is not in the document]
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: How can its size be included there if it's not constant but an arbitrary expression? I can see you have a test for it (`new int[x]` != `new int[z]`), and that it appears to work, but how?

[the remark this answers is not in the document]
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll]: Oh, now I see why it's working. You added a commit to fix it.

**Comment by jbj** [comment]
#116 has been merged now, so I assume you want to sync up with that.

**Comment by felicity-semmle** [comment]
Should this be mentioned in the analysis change notes for 1.18?

**Comment by jbj** [comment]
@felicity-semmle Yes, if it makes it into 1.18.

File: change-notes/1.18/analysis-cpp.md [hunk 14]
```diff
@@ -33,4 +33,4 @@
  
 ## Changes to QL libraries
 
-* *Series of bullet points*
+* Added a hash consing library for structural comparison of expressions.
```
Comment (jbj) [inline change-notes/1.18/analysis-cpp.md]: Please refer readers to where they can find the library. Maybe instead write "Added a new library `semmle.code.cpp.valuenumbering.HashCons` for structural comparison of expressions."

Also, this file is in conflict.

**Comment by jbj** [comment]
I've taken this out of the 1.18 milestone as it keeps slipping, and I want to make sure the team focus is on testing and stabilising 1.18 rather than adding this feature. I still hope we can merge this soon, but testing 1.18 should take priority over this for both me and @rdmarsh2.

Please rebase to `master` and move the change note into the 1.19 change notes file (create if needed).

**Comment by rdmarsh2** [comment]
Moved change notes and added the LGTM import. I believe I've addressed all the review comments.
